ipn/ipnlocal/local.go

```diff
@@ -5990,3 +5993,90 @@ func mayDeref[T any](p *T) (v T) {
 	}
 	return *p
 }
+
+// productionRNG is used for randomly picking an exit node in SuggestDERPExitNode.
+var productionRNG = rand.New(rand.NewSource(rand.Int63()))
+
+func (b *LocalBackend) suggestExitNodeEnabled() bool {
+	return b.ControlKnobs().SuggestExitNode.Load()
+}
+
+const derpPrefix = "127.3.3.40:"
+
+// SuggestDERPExitNode returns a tailcfg.StableNodeID of a suggested exit node given the local backend's netmap and last report.
+func (b *LocalBackend) SuggestDERPExitNode() (tailcfg.StableNodeID, error) {
+	b.mu.Lock()
+	lastReport := b.MagicConn().GetLastNetcheckReport(b.ctx)
+	netMap := b.netMap
+	b.mu.Unlock()
+	if b.suggestExitNodeEnabled() {
+		return suggestDERPExitNode(lastReport, netMap, productionRNG)
+	} else {
+		return "", fmt.Errorf("Unable to choose exit node")
+	}
+}
+
+func suggestDERPExitNode(lastReport *netcheck.Report, netMap *netmap.NetworkMap, rng *rand.Rand) (tailcfg.StableNodeID, error) {
+	peers := netMap.Peers
+	var preferredExitNodeID tailcfg.StableNodeID
+	peerRegionMap := make(map[int][]tailcfg.NodeView)
```

Comment:
This is not as important, but we do have a reasonable guess as to the number of entries in the map, so might as well use it to maybe reduce allocations.

```suggestion
	peerRegionMap := make(map[int][]tailcfg.NodeView, len(netMap.DERPMap.Regions)
```